lib/graphql/query/serial_execution/field_resolution.rb

```diff
@@ -32,7 +32,10 @@ def result
         # continue by "finishing" the value, eg. executing sub-fields or coercing values
         def get_finished_value(raw_value)
           raise raw_value if raw_value.instance_of?(GraphQL::ExecutionError)
-          return nil if raw_value.nil?
+          if raw_value.nil?
+            raise GraphQL::ExecutionError, "Type mismatch resolving '#{field.name}'" if field.type.kind.non_null?
```

Comment:
OK yes, I'd already kind of wandered down the "write an `InterfaceResolution` class" path, and splitting `BaseResolution#result` in two was the extra piece I needed.

I think I'm OK with this PR as-is now; permitting coercion of nil for scalars is worthy of its own PR.